alembic/versions/20230512_5a425ebe026c_migrate_millenium_apis_to_post.py

```diff
@@ -0,0 +1,98 @@
+"""Migrate millenium APIs to POST
+
+Revision ID: 5a425ebe026c
+Revises: f9985f6b7767
+Create Date: 2023-05-12 08:36:16.603825+00:00
+
+"""
+import logging
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "5a425ebe026c"
+down_revision = "f9985f6b7767"
+branch_labels = None
+depends_on = None
+
+
+log = logging.getLogger(f"palace.migration.{revision}")
+log.setLevel(logging.INFO)
+log.disabled = False
+
+KEY = "use_post_requests"
+
+
+def upgrade() -> None:
+    """Set 'use_post_requests' to 'true' for 'api.millenium' integrations.
+    But only those that are for the following urls
+    - thepalaceproject.org
+    - localhost
+    """
+    conn = op.get_bind()
+    # Find the relevant external integrations
+    result_ids = conn.execute(
+        "SELECT id FROM externalintegrations where protocol='api.millenium_patron'"
+    )
+
+    # Query to pull specific config values
+    config_query = "SELECT value from configurationsettings where external_integration_id={integration_id} and key='{key}'"
+
+    # For each millenium integration found
+    for [integration_id] in result_ids or []:
+        # Pull the URL setting
+        config_results = conn.execute(
+            config_query.format(integration_id=integration_id, key="url")
+        )
+        url_results = list(config_results)
+        if config_results and len(url_results) > 0:
+            url = url_results[0][0]
+        else:
+            log.info(f"No URL found for integration: {integration_id}")
+            continue
+
+        # Check if it is something we want to change at all
+        if not ("thepalaceproject.org" in url or "localhost" in url):
```

Comment:
Yes this is true :point_up: 
I'm not sure it's relevant though since the subset of urls used in this configuration is quite limited.

Reply:
It seems dismissable to me. The input is kind of trusted given that we configured it in the first place.

Reply:
Maybe we could just run this through `urlparse` and match on hostname, so that we don't have to dismiss the error? 

I'd also like to be a bit more specific about what urls get migrated. We should match on `vlc.thepalaceproject.org` for production, but we should also match the dev / qa servers as well. I think matching on `vlc.*.palaceproject.io` would cover theses cases for `vlc.dev.palaceproject.io`, `vlc.staging.palaceproject.io`, etc.

Reply:
I was addressing the error from the perspective of "this isn't a security issue". I think the error is saying "your security might be dependent on sanitizing this URL, and you're not doing what you appear to think you're doing". I don't think this is a security issue, because these URLs aren't coming from an untrusted source.

Being more specific about the URLs sounds good though. I wasn't aware we had other domain names that wouldn't be matched by this.